// gatsby-plugin-gdpr-cookies/gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _react = _interopRequireDefault(require("react"));

var _commonTags = require("common-tags");

var _merge = _interopRequireDefault(require("lodash/merge"));

var _validTrackingId = require("./valid-tracking-id");

var _defaultOptions = _interopRequireDefault(require("./default-options"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

var generateGTM = function generateGTM(id, dataLayerName, environmentParamStr) {
  return (0, _commonTags.stripIndent)(_templateObject || (_templateObject = (0, _taggedTemplateLiteralLoose2.default)(["\n  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':\n  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],\n  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=\n  'https://www.googletagmanager.com/gtm.js?id='+i+dl+'", "';f.parentNode.insertBefore(j,f);\n  })(window,document,'script','", "', '", "');"])), environmentParamStr, dataLayerName, id);
};

var generateGTMIframe = function generateGTMIframe(id, environmentParamStr) {
  return (0, _commonTags.oneLine)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteralLoose2.default)(["<iframe src=\"https://www.googletagmanager.com/ns.html?id=", "", "\" height=\"0\" width=\"0\" style=\"display: none; visibility: hidden\"></iframe>"])), id, environmentParamStr);
};

var generateGTMDefaultDataLayer = function generateGTMDefaultDataLayer(dataLayer, dataLayerName, reporter) {
  var result = "window." + dataLayerName + " = window." + dataLayerName + " || [];";

  if (dataLayer.type === "function") {
    result += "window." + dataLayerName + ".push((" + dataLayer.value + ")());";
  } else {
    if (dataLayer.type !== "object" || dataLayer.value.constructor !== Object) {
      reporter.panic("Oops the plugin option \"defaultDataLayer\" should be a plain object. \"" + dataLayer + "\" is not valid.");
    }

    result += "window." + dataLayerName + ".push(" + JSON.stringify(dataLayer.value) + ");";
  }

  return (0, _commonTags.stripIndent)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteralLoose2.default)(["", ""])), result);
}; // add scripts


exports.onRenderBody = function (_ref, pluginOptions) {
  var setHeadComponents = _ref.setHeadComponents,
      setPreBodyComponents = _ref.setPreBodyComponents,
      reporter = _ref.reporter;

  if (pluginOptions === void 0) {
    pluginOptions = {};
  }

  var currentEnvironment = process.env.ENV || process.env.NODE_ENV || "development";
  var options = (0, _merge.default)(_defaultOptions.default, pluginOptions);
  var headComponents = [];
  var preBodyComponents = []; // add pre connection to analytics

  headComponents.push( /*#__PURE__*/_react.default.createElement("link", {
    rel: "preconnect dns-prefetch",
    key: "preconnect-google-analytics",
    href: "https://www.google-analytics.com"
  })); // facebook pixel

  if (options.environments.includes(currentEnvironment) && (0, _validTrackingId.validFbPixelId)(options)) {
    headComponents.push( /*#__PURE__*/_react.default.createElement("script", {
      key: "gatsby-plugin-gdpr-cookies-facebook-pixel",
      dangerouslySetInnerHTML: {
        __html: (0, _commonTags.stripIndent)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteralLoose2.default)(["\n            !function(f,b,e,v,n,t,s)\n            {if(f.fbq)return;n=f.fbq=function(){n.callMethod?\n            n.callMethod.apply(n,arguments):n.queue.push(arguments)};\n            if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';\n            n.queue=[];t=b.createElement(e);t.async=!0;\n            t.src=v;s=b.getElementsByTagName(e)[0];\n            s.parentNode.insertBefore(t,s)}(window, document,'script',\n            'https://connect.facebook.net/en_US/fbevents.js');\n          "])))
      }
    }));
  }

  var googleAnalytics = options.googleAnalytics; // Hey window['ga-disable-MEASUREMENT_ID'] = true;

  headComponents.push( /*#__PURE__*/_react.default.createElement("script", {
    key: "gatsby-plugin-gdpr-cookies-google-gtag-disable",
    dangerouslySetInnerHTML: {
      __html: (0, _commonTags.stripIndent)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteralLoose2.default)(["\n          window['ga-disable-", "'] = true\n        "])), googleAnalytics.trackingId)
    }
  }));
  headComponents.push( /*#__PURE__*/_react.default.createElement("script", {
    key: "gatsby-plugin-gdpr-cookies-google-gtag-script",
    async: true,
    src: "https://www.googletagmanager.com/gtag/js?id=" + googleAnalytics.trackingId
  })); // google tag manager

  if (options.environments.includes(currentEnvironment) && (0, _validTrackingId.validGTMTrackingId)(options)) {
    var googleTagManager = options.googleTagManager;
    var environmentParamStr = "";

    if (googleTagManager.gtmAuth && googleTagManager.gtmPreview) {
      environmentParamStr = (0, _commonTags.oneLine)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteralLoose2.default)(["\n        &gtm_auth=", "&gtm_preview=", "&gtm_cookies_win=x\n      "])), googleTagManager.gtmAuth, googleTagManager.gtmPreview);
    }

    var defaultDataLayerCode = "";

    if (googleTagManager.defaultDataLayer) {
      defaultDataLayerCode = generateGTMDefaultDataLayer(googleTagManager.defaultDataLayer, googleTagManager.dataLayerName, reporter);
    }

    headComponents.push( /*#__PURE__*/_react.default.createElement("script", {
      key: "gatsby-plugin-gdpr-cookies-google-tagmanager",
      dangerouslySetInnerHTML: {
        __html: (0, _commonTags.oneLine)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteralLoose2.default)(["\n            ", "\n            ", "\n          "])), defaultDataLayerCode, generateGTM(googleTagManager.trackingId, googleTagManager.dataLayerName, environmentParamStr))
      }
    }));
    preBodyComponents.push( /*#__PURE__*/_react.default.createElement("noscript", {
      key: "gatsby-plugin-gdpr-cookies-google-tagmanager",
      dangerouslySetInnerHTML: {
        __html: generateGTMIframe(googleTagManager.trackingId, environmentParamStr)
      }
    }));
  } // add scripts


  if (headComponents.length) {
    setHeadComponents(headComponents);
  }

  if (preBodyComponents.length) {
    setPreBodyComponents(preBodyComponents);
  }
};
